Let the compiler generate more comparison operators in WebCore

cdumez opened the pull request:
#### f1d8394cd5374197cb332d34d5fe9e46f80a7e1c
<pre>
Let the compiler generate more comparison operators in WebCore
<a href="https://bugs.webkit.org/show_bug.cgi?id=261076">https://bugs.webkit.org/show_bug.cgi?id=261076</a>

Reviewed by Darin Adler.

Let the compiler generate more comparison operators in WebCore now that we
support C++20.

* Source/WebCore/platform/graphics/ColorInterpolationMethod.h:
(WebCore::operator==): Deleted.
* Source/WebCore/platform/graphics/ColorMatrix.h:
(WebCore::operator==): Deleted.
* Source/WebCore/platform/graphics/DecodingOptions.h:
(WebCore::DecodingOptions::operator== const): Deleted.
* Source/WebCore/platform/graphics/FloatPoint.h:
(WebCore::operator==): Deleted.
* Source/WebCore/platform/graphics/FloatPoint3D.h:
(WebCore::operator==): Deleted.
* Source/WebCore/platform/graphics/FloatQuad.h:
(WebCore::operator==): Deleted.
* Source/WebCore/platform/graphics/FloatRect.h:
(WebCore::operator==): Deleted.
* Source/WebCore/platform/graphics/FloatRoundedRect.h:
(WebCore::operator==): Deleted.
* Source/WebCore/platform/graphics/FloatSize.h:
(WebCore::operator==): Deleted.
* Source/WebCore/platform/graphics/FontCache.cpp:
(WebCore::operator==): Deleted.
* Source/WebCore/platform/graphics/FontCascadeCache.cpp:
* Source/WebCore/platform/graphics/FontCascadeCache.h:
* Source/WebCore/platform/graphics/FontCascadeDescription.h:
(WebCore::FontCascadeDescription::operator== const):
* Source/WebCore/platform/graphics/FontDescription.h:
(WebCore::FontDescription::operator== const): Deleted.
* Source/WebCore/platform/graphics/FontSelectionAlgorithm.h:
(WebCore::operator==): Deleted.
(WebCore::FontSelectionRange::operator== const): Deleted.
* Source/WebCore/platform/graphics/FontSizeAdjust.h:
(WebCore::FontSizeAdjust::operator== const): Deleted.
* Source/WebCore/platform/graphics/FontTaggedSettings.h:
(WebCore::= const): Deleted.
(WebCore::FontTaggedSettings::operator== const): Deleted.
* Source/WebCore/platform/graphics/FourCC.h:
(WebCore::operator==): Deleted.
* Source/WebCore/platform/graphics/GraphicsStyle.h:
* Source/WebCore/platform/graphics/GraphicsTypes.h:
(WebCore::operator==): Deleted.
* Source/WebCore/platform/graphics/ImageOrientation.h:
* Source/WebCore/platform/graphics/ImageSource.h:
* Source/WebCore/platform/graphics/IntPoint.h:
(WebCore::operator==): Deleted.
* Source/WebCore/platform/graphics/IntRect.h:
(WebCore::operator==): Deleted.
* Source/WebCore/platform/graphics/IntSize.h:
(WebCore::operator==): Deleted.
* Source/WebCore/platform/graphics/LayoutPoint.h:
(WebCore::operator==): Deleted.
* Source/WebCore/platform/graphics/LayoutRect.h:
(WebCore::operator==): Deleted.
* Source/WebCore/platform/graphics/LayoutSize.h:
(WebCore::operator==): Deleted.
* Source/WebCore/platform/graphics/MediaUsageInfo.h:
(WebCore::MediaUsageInfo::operator== const): Deleted.
* Source/WebCore/platform/graphics/PathTraversalState.cpp:
(WebCore::QuadraticBezier::operator== const): Deleted.
(WebCore::CubicBezier::operator== const): Deleted.
* Source/WebCore/platform/graphics/PlatformAudioTrackConfiguration.h:
(WebCore::operator==): Deleted.
* Source/WebCore/platform/graphics/PlatformTimeRanges.cpp:
(WebCore::PlatformTimeRanges::operator== const): Deleted.
* Source/WebCore/platform/graphics/PlatformTimeRanges.h:
* Source/WebCore/platform/graphics/PlatformTrackConfiguration.h:
* Source/WebCore/platform/graphics/PlatformVideoColorSpace.h:
(WebCore::operator==): Deleted.
* Source/WebCore/platform/graphics/PlatformVideoTrackConfiguration.h:
(WebCore::operator==): Deleted.
* Source/WebCore/platform/graphics/Region.h:
* Source/WebCore/platform/graphics/RoundedRect.h:
(WebCore::operator==): Deleted.
* Source/WebCore/platform/graphics/ScreenDataOverrides.h:
(WebCore::operator==): Deleted.
* Source/WebCore/platform/graphics/SourceBrush.h:
* Source/WebCore/platform/graphics/TabSize.h:
(WebCore::operator==): Deleted.
* Source/WebCore/platform/graphics/TextBoxIterator.h:
(WebCore::TextBoxIterator::operator== const): Deleted.
* Source/WebCore/platform/graphics/avfoundation/objc/CDMInstanceFairPlayStreamingAVFObjC.h:
* Source/WebCore/platform/graphics/avfoundation/objc/InbandChapterTrackPrivateAVFObjC.h:
(WebCore::InbandChapterTrackPrivateAVFObjC::ChapterData::operator== const): Deleted.
* Source/WebCore/platform/graphics/cocoa/SystemFontDatabaseCoreText.h:
(WebCore::SystemFontDatabaseCoreText::CascadeListParameters::operator== const): Deleted.
* Source/WebCore/platform/graphics/cv/GraphicsContextGLCVCocoa.cpp:
(WebCore::GraphicsContextGLCVCocoa::TextureContent::operator== const): Deleted.
* Source/WebCore/platform/graphics/cv/GraphicsContextGLCVCocoa.h:
* Source/WebCore/platform/graphics/cv/ImageRotationSessionVT.h:
(WebCore::operator==): Deleted.
* Source/WebCore/platform/graphics/transforms/AffineTransform.h:
(WebCore::AffineTransform::operator== const): Deleted.
* Source/WebCore/platform/graphics/transforms/TransformationMatrix.h:
(WebCore::TransformationMatrix::Decomposed2Type::operator== const): Deleted.
(WebCore::TransformationMatrix::Decomposed4Type::operator== const): Deleted.
* Source/WebCore/platform/network/FormData.h:
(WebCore::FormDataElement::EncodedFileData::operator== const): Deleted.
(WebCore::FormDataElement::EncodedBlobData::operator== const): Deleted.
* Source/WebCore/platform/network/HTTPHeaderMap.h:
(WebCore::HTTPHeaderMap::CommonHeader::operator== const): Deleted.
(WebCore::HTTPHeaderMap::UncommonHeader::operator== const): Deleted.
* Source/WebCore/platform/network/cf/CertificateInfo.h:
(WebCore::CertificateInfo::operator== const): Deleted.
* Source/WebCore/platform/text/StringWithDirection.h:
(WebCore::operator==): Deleted.
* Source/WebCore/platform/text/TextFlags.h:
(WebCore::ExpansionBehavior::operator== const): Deleted.
(WebCore::FontVariantAlternatesValues::operator== const): Deleted.
(WebCore::FontVariantAlternates::operator== const): Deleted.
(WebCore::FontVariantSettings::operator== const): Deleted.
* Source/WebCore/platform/text/TextSpacing.h:
(WebCore::TextSpacingTrim::isAuto const):
(WebCore::TextSpacingTrim::isSpaceAll const):
(WebCore::TextAutospace::isAuto const):
(WebCore::TextAutospace::isNoAutospace const):
(WebCore::TextSpacingTrim::operator== const): Deleted.
(WebCore::TextAutospace::operator== const): Deleted.
* Source/WebCore/plugins/PluginData.h:
(WebCore::operator==): Deleted.

Canonical link: <a href="https://commits.webkit.org/267597@main">https://commits.webkit.org/267597@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a65099517d5213c8072d5276e2d3062a0e9caed8

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/17120 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/17445 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/17935 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/18906 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/16018 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/17312 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/20719 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/17587 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/18218 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/17323 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/17673 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/14857 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/19722 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/14913 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/15548 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/22239 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/15912 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/15715 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/20056 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/16306 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/13829 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/15461 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4086 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/19827 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/16137 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/platform/graphics/FontCascadeDescription.h:
```diff
@@ -160,7 +160,7 @@ class FontCascadeDescription : public FontDescription {
 inline bool FontCascadeDescription::operator==(const FontCascadeDescription& other) const
 {
-    return FontDescription::operator==(other)
+    return static_cast<FontDescription>(*this) == static_cast<FontDescription>(other)
```
How did you decide to cast to `FontDescription` rather than to `const FontDescription&`?

cdumez replied:
That's my mistake, I'll fix.

darinadler commented on Source/WebCore/platform/graphics/FontSelectionAlgorithm.h:
```diff
@@ -403,11 +399,6 @@ inline TextStream& operator<<(TextStream& ts, const std::optional<FontSelectionV
 
-inline bool operator==(const FontSelectionRequest& a, const FontSelectionRequest& b)
-{
-    return a.tied() == b.tied();
```
Maybe we should delete the `tied` function too and just move the code into the `add(Hasher&)` function?

darinadler commented on Source/WebCore/platform/graphics/FontSelectionAlgorithm.h:
```diff
@@ -433,11 +426,6 @@ struct FontSelectionCapabilities {
 
-constexpr bool operator==(const FontSelectionCapabilities& a, const FontSelectionCapabilities& b)
-{
-    return a.tied() == b.tied();
```
Maybe we should delete the `tied` function too and just move the code into the `add(Hasher&)` function?

darinadler commented on Source/WebCore/platform/graphics/FontSelectionAlgorithm.h:
```diff
@@ -519,11 +509,6 @@ std::optional<FontSelectionSpecifiedCapabilities> FontSelectionSpecifiedCapabili
 
-constexpr bool operator==(const FontSelectionSpecifiedCapabilities& a, const FontSelectionSpecifiedCapabilities& b)
-{
-    return a.tied() == b.tied();
```
Maybe we should delete the `tied` function too and just move the code into the `add(Hasher&)` function?

darinadler commented on Source/WebCore/platform/graphics/ImageOrientation.h:
```diff
@@ -75,7 +75,7 @@ struct ImageOrientation {
     constexpr operator int() const { return static_cast<int>(m_orientation); }
-    friend constexpr bool operator==(const ImageOrientation& lhs, const ImageOrientation& rhs) { return lhs.m_orientation == rhs.m_orientation; }
+    friend constexpr bool operator==(const ImageOrientation&, const ImageOrientation&) = default;
     friend constexpr bool operator==(const ImageOrientation& lhs, const Orientation& rhs) { return lhs.m_orientation == rhs; }
```
Is this overload needed? Does removing it cause a compilation failure?

cdumez replied:
Doesn't seem needed, will remove.

darinadler commented on Source/WebCore/platform/graphics/Region.h:
```diff
@@ -79,6 +79,8 @@ class Region {
         int y { 0 };
         size_t segmentIndex { 0 };
+
+        friend bool operator==(const Region::Span&, const Region::Span&) = default;
```
Can’t we just write `Span` here instead of `Region::Span`?

darinadler commented on Source/WebCore/platform/text/TextSpacing.h:
```diff
@@ -30,19 +30,15 @@
 namespace WebCore {
 
 struct TextSpacingTrim {
+    enum class TrimType: uint8_t {
```
Space before colon? Use `bool` instead of `uint8_t`?

darinadler commented on Source/WebCore/platform/text/TextSpacing.h:
```diff
@@ -58,18 +54,15 @@ inline WTF::TextStream& operator<<(WTF::TextStream& ts, const TextSpacingTrim& v
-    Auto = 0,
-    NoAutospace // equivalent to None in text-spacing shorthand
-};
+    enum class TextAutospaceType: uint8_t {
```
Space before colon? Use `bool` instead of `uint8_t`?

cdumez committed:
Let the compiler generate more comparison operators in WebCore
https://bugs.webkit.org/show_bug.cgi?id=261076

Reviewed by Darin Adler.

Let the compiler generate more comparison operators in WebCore now that we
support C++20.

* Source/WebCore/platform/graphics/ColorInterpolationMethod.h:
(WebCore::operator==): Deleted.
* Source/WebCore/platform/graphics/ColorMatrix.h:
(WebCore::operator==): Deleted.
* Source/WebCore/platform/graphics/DecodingOptions.h:
(WebCore::DecodingOptions::operator== const): Deleted.
* Source/WebCore/platform/graphics/FloatPoint.h:
(WebCore::operator==): Deleted.
* Source/WebCore/platform/graphics/FloatPoint3D.h:
(WebCore::operator==): Deleted.
* Source/WebCore/platform/graphics/FloatQuad.h:
(WebCore::operator==): Deleted.
* Source/WebCore/platform/graphics/FloatRect.h:
(WebCore::operator==): Deleted.
* Source/WebCore/platform/graphics/FloatRoundedRect.h:
(WebCore::operator==): Deleted.
* Source/WebCore/platform/graphics/FloatSize.h:
(WebCore::operator==): Deleted.
* Source/WebCore/platform/graphics/FontCache.cpp:
(WebCore::operator==): Deleted.
* Source/WebCore/platform/graphics/FontCascadeCache.cpp:
* Source/WebCore/platform/graphics/FontCascadeCache.h:
* Source/WebCore/platform/graphics/FontCascadeDescription.h:
(WebCore::FontCascadeDescription::operator== const):
* Source/WebCore/platform/graphics/FontDescription.h:
(WebCore::FontDescription::operator== const): Deleted.
* Source/WebCore/platform/graphics/FontSelectionAlgorithm.h:
(WebCore::operator==): Deleted.
(WebCore::FontSelectionRange::operator== const): Deleted.
* Source/WebCore/platform/graphics/FontSizeAdjust.h:
(WebCore::FontSizeAdjust::operator== const): Deleted.
* Source/WebCore/platform/graphics/FontTaggedSettings.h:
(WebCore::= const): Deleted.
(WebCore::FontTaggedSettings::operator== const): Deleted.
* Source/WebCore/platform/graphics/FourCC.h:
(WebCore::operator==): Deleted.
* Source/WebCore/platform/graphics/GraphicsStyle.h:
* Source/WebCore/platform/graphics/GraphicsTypes.h:
(WebCore::operator==): Deleted.
* Source/WebCore/platform/graphics/ImageOrientation.h:
* Source/WebCore/platform/graphics/ImageSource.h:
* Source/WebCore/platform/graphics/IntPoint.h:
(WebCore::operator==): Deleted.
* Source/WebCore/platform/graphics/IntRect.h:
(WebCore::operator==): Deleted.
* Source/WebCore/platform/graphics/IntSize.h:
(WebCore::operator==): Deleted.
* Source/WebCore/platform/graphics/LayoutPoint.h:
(WebCore::operator==): Deleted.
* Source/WebCore/platform/graphics/LayoutRect.h:
(WebCore::operator==): Deleted.
* Source/WebCore/platform/graphics/LayoutSize.h:
(WebCore::operator==): Deleted.
* Source/WebCore/platform/graphics/MediaUsageInfo.h:
(WebCore::MediaUsageInfo::operator== const): Deleted.
* Source/WebCore/platform/graphics/PathTraversalState.cpp:
(WebCore::QuadraticBezier::operator== const): Deleted.
(WebCore::CubicBezier::operator== const): Deleted.
* Source/WebCore/platform/graphics/PlatformAudioTrackConfiguration.h:
(WebCore::operator==): Deleted.
* Source/WebCore/platform/graphics/PlatformTimeRanges.cpp:
(WebCore::PlatformTimeRanges::operator== const): Deleted.
* Source/WebCore/platform/graphics/PlatformTimeRanges.h:
* Source/WebCore/platform/graphics/PlatformTrackConfiguration.h:
* Source/WebCore/platform/graphics/PlatformVideoColorSpace.h:
(WebCore::operator==): Deleted.
* Source/WebCore/platform/graphics/PlatformVideoTrackConfiguration.h:
(WebCore::operator==): Deleted.
* Source/WebCore/platform/graphics/Region.h:
* Source/WebCore/platform/graphics/RoundedRect.h:
(WebCore::operator==): Deleted.
* Source/WebCore/platform/graphics/ScreenDataOverrides.h:
(WebCore::operator==): Deleted.
* Source/WebCore/platform/graphics/SourceBrush.h:
* Source/WebCore/platform/graphics/TabSize.h:
(WebCore::operator==): Deleted.
* Source/WebCore/platform/graphics/TextBoxIterator.h:
(WebCore::TextBoxIterator::operator== const): Deleted.
* Source/WebCore/platform/graphics/avfoundation/objc/CDMInstanceFairPlayStreamingAVFObjC.h:
* Source/WebCore/platform/graphics/avfoundation/objc/InbandChapterTrackPrivateAVFObjC.h:
(WebCore::InbandChapterTrackPrivateAVFObjC::ChapterData::operator== const): Deleted.
* Source/WebCore/platform/graphics/cocoa/SystemFontDatabaseCoreText.h:
(WebCore::SystemFontDatabaseCoreText::CascadeListParameters::operator== const): Deleted.
* Source/WebCore/platform/graphics/cv/GraphicsContextGLCVCocoa.cpp:
(WebCore::GraphicsContextGLCVCocoa::TextureContent::operator== const): Deleted.
* Source/WebCore/platform/graphics/cv/GraphicsContextGLCVCocoa.h:
* Source/WebCore/platform/graphics/cv/ImageRotationSessionVT.h:
(WebCore::operator==): Deleted.
* Source/WebCore/platform/graphics/transforms/AffineTransform.h:
(WebCore::AffineTransform::operator== const): Deleted.
* Source/WebCore/platform/graphics/transforms/TransformationMatrix.h:
(WebCore::TransformationMatrix::Decomposed2Type::operator== const): Deleted.
(WebCore::TransformationMatrix::Decomposed4Type::operator== const): Deleted.
* Source/WebCore/platform/network/FormData.h:
(WebCore::FormDataElement::EncodedFileData::operator== const): Deleted.
(WebCore::FormDataElement::EncodedBlobData::operator== const): Deleted.
* Source/WebCore/platform/network/HTTPHeaderMap.h:
(WebCore::HTTPHeaderMap::CommonHeader::operator== const): Deleted.
(WebCore::HTTPHeaderMap::UncommonHeader::operator== const): Deleted.
* Source/WebCore/platform/network/cf/CertificateInfo.h:
(WebCore::CertificateInfo::operator== const): Deleted.
* Source/WebCore/platform/text/StringWithDirection.h:
(WebCore::operator==): Deleted.
* Source/WebCore/platform/text/TextFlags.h:
(WebCore::ExpansionBehavior::operator== const): Deleted.
(WebCore::FontVariantAlternatesValues::operator== const): Deleted.
(WebCore::FontVariantAlternates::operator== const): Deleted.
(WebCore::FontVariantSettings::operator== const): Deleted.
* Source/WebCore/platform/text/TextSpacing.h:
(WebCore::TextSpacingTrim::isAuto const):
(WebCore::TextSpacingTrim::isSpaceAll const):
(WebCore::TextAutospace::isAuto const):
(WebCore::TextAutospace::isNoAutospace const):
(WebCore::TextSpacingTrim::operator== const): Deleted.
(WebCore::TextAutospace::operator== const): Deleted.
* Source/WebCore/plugins/PluginData.h:
(WebCore::operator==): Deleted.

Canonical link: https://commits.webkit.org/267597@main
